### ISSUE TITLE
dealii-9.0.0.dmg OSX and "illegal instruction 4"

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#7175]
@luca-heltai we have had several reports from users of the .dmg that they can not run code (ASPECT) because they trigger "illegal instruction 4" errors. I assume this is because some of the dependencies are getting compiled using machine specific optimizations.

Also see https://github.com/dealii/dealii/issues/4109#issuecomment-294268261

and https://groups.google.com/forum/#!topic/spack/-tlNZhXbKOo

Do you have an idea what to do about this?

**Comment by tjhei** [comment]
I can now reproduce this as I have access to a 2012 iMac (i5-3470). Newest xcode and OS and step-40 crashes with this error.

**Comment by tjhei** [comment]
And the error is in ``libopenbl_haswell``...

**Comment by luca-heltai** [comment]
I have no idea what is happening, nor if we can avoid this. @davydden may know... is there a way to stop spack from using hardware specific optimization when compiling? If so, I could try to re-build everything without optimization.

**Comment by davydden** [comment]
> is there a way to stop spack from using hardware specific optimization when compiling?

Spack does not unconditionally add any optimization flags. But some packages might have those in their recepies. 

There is also an option to set complex flags in compiler wrappers, but I think it is used exactly for the opposite reasons - optimize all builds to given hardware

**Comment by bangerth** [comment]
Do any of the recipes use `-march=native`? You may have to explicitly set `-march=i686` or something like this with minimal requirements.

Though the link dependency on `libopenbl_haswell` is interesting. Do you know (using `otool`) where this comes from? I.e., which dependency library links against `libopenbl_haswell`?

**Comment by davydden** [comment]
> Do you know (using otool) where this comes from?

I think that's where we should start from... Otherwise deal.II has far too many link dependencies in the graph to randomly check each of them.

**Comment by tjhei** [comment]
This is the ``openblas`` library, which comes in from petsc->hypre->openblas for example:
```
$ ./bin/spack spec petsc
petsc@3.9.0%clang@9.1.0-apple+boost clanguage=C ~complex~debug+double+hdf5+hypre~int64+metis+mpi+mumps+shared~suite-sparse+superlu-dist~trilinos arch=darwin-highsierra-x86_64 
    ^hypre@2.14.0%clang@9.1.0-apple~int64~internal-superlu+mpi~shared arch=darwin-highsierra-x86_64 
        ^openblas@0.2.20%clang@9.1.0-apple cpu_target= ~ilp64 patches=47cfa7a952ac7b2e4632c73ae199d69fb54490627b66a62c681e21019c4ddc9d +pic+shared threads=none ~virtual_machine arch=darwin-highsierra-x86_64 
...
```
but there are other dependencies, for example trilinos->hypre->openblas, or suite-sparse->openblas, or superlu-dist->openblas.

**Comment by davydden** [comment]
Spack does not set any flags in openblas: 
https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/openblas/package.py

Maybe one can set -march=xyz in compiler wrappers and this will be used for all packages by construction.

**Comment by davydden** [comment]
A better way is to set cpu_target variant in openblas

**Comment by luca-heltai** [comment]
On it.

**Comment by luca-heltai** [comment]
I have just uploaded a new version of the package:

https://github.com/dealii/dealii/releases/download/v9.0.0/dealii-9.0.0.dmg

Can any of you try if this work on older machines/different setups?

**Comment by tjhei** [comment]
> Can any of you try if this work on older machines/different setups?

I will. Wait, did you just replace the existing .dmg? That will cause some issues, for example the mirror now offers a different file under the same name. Can you give it a different name at least?

**Comment by tjhei** [comment]
> Can any of you try

I am getting:
```
bash-3.2$ ./step-40
dyld: Library not loaded: libiomp5.dylib
  Referenced from: /Applications/deal.II-9.0.0.app/Contents/Resources/spack/opt/spack/darwin-highsierra-x86_64/clang-6.0.0/assimp-4.0.1-pdlrx32676yct2pztierxgvynk5gx277/lib/libassimp.4.dylib
  Reason: image not found
Abort trap: 6
```

The file ``./Contents/Resources/spack/clang/lib/libiomp5.dylib`` is not found for some reason.

**Comment by luca-heltai** [comment]
do you have the same problem if you run the existing executables in  /Applications/deal.II-9.0.0.app/Contents/Resources/examples/step-40?

**Comment by luca-heltai** [comment]
(I renamed the images to `deal.II-9.0.0-clang-6.0.dmg`). I'll upload back the other images from the Mirror.

**Comment by tjhei** [comment]
> do you have the same problem if you run the existing executables in /Applications/deal.II-9.0.0.app/Contents/Resources/examples/step-40?

yes (and this is from the deal.II terminal).

**Comment by luca-heltai** [comment]
I have no idea how to fix this... 

can you try to make package with your mac?

**Comment by tjhei** [comment]
> (I renamed the images

thanks! I wouldn't mind if we version these .dmg files in the future, maybe as 9.0.0-r1, r2, etc.

**Comment by tjhei** [comment]
This should be fixed with the latest 9.2 dmg file.
